### ISSUE TITLE
`startEvent` Client API in `bpmn-client`

### ISSUE THREAD

**Issue by MaheshkumarSundaram** [bpmnServer/bpmn-server#179]
Hi @ralphhanna,

Please add `api.engine.startEvent` in `bpmn-client` and also in `src/scripts/cli.ts` in the webapp. Thanks.

**Comment by ralphhanna** [comment]
Done, please check and close
Thanks

**Comment by MaheshkumarSundaram** [comment]
Hi @ralphhanna,

I have a completed process like below:
![image](https://github.com/bpmnServer/bpmn-server/assets/55505611/ce6f4767-12a2-40c5-869e-c04f3d5099d9)

I am using `bpmn-client` to call start the above completed process again with its starting point from `RecoveryStart`.
![image](https://github.com/bpmnServer/bpmn-server/assets/55505611/9f944807-c985-4c3d-8401-bc1aad399865)

But it results in error. What am I doing wrong?

**Comment by MaheshkumarSundaram** [comment]
@ralphhanna,

API needs to be changed in `bpmn-client` to accomodate optional params - `userName` & `options` for `startEvent`.

But I tried using the `src/scripts/cli.ts` in `bpmn-web` by adding `restart` option:

```js
let response = await server.engine.startEvent(instanceId, nodeId, data, null, { restart: 'true' });
```
But still the invoke fails and I don't see any console errors in the server.

![image](https://github.com/bpmnServer/bpmn-server/assets/55505611/c37de79e-1c63-489a-9821-85c0d50185fb)

Attaching my BPMN file if you would like to test it on your end (I have my `enableSavePoints` disabled as its not required for `startEvent`).

<details>
  <summary><bold>BPMN File</bold></summary>

  ```xml
 <?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn">
  <bpmn:process id="Process_1" isExecutable="true">
    <bpmn:startEvent id="StartEvent_1" name="Start">
      <bpmn:extensionElements>
        <camunda:properties>
          <camunda:property name="source" value="MSNODE3" />
          <camunda:property name="protocol" value="H&#38;n" />
          <camunda:property name="modality" value="CT" />
        </camunda:properties>
      </bpmn:extensionElements>
      <bpmn:outgoing>Flow_1ky5rfq</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:sequenceFlow id="Flow_1ky5rfq" sourceRef="StartEvent_1" targetRef="Activity_1yfed52" />
    <bpmn:serviceTask id="Activity_1yfed52" name="ST1" camunda:delegateExpression="Dummy">
      <bpmn:extensionElements>
        <camunda:inputOutput>
          <camunda:outputParameter name="myVar">$[{ 'dest': '2', 'mod': 'RTSTRUCT'}, { 'dest': '1', 'mod': 'CT'}]</camunda:outputParameter>
        </camunda:inputOutput>
      </bpmn:extensionElements>
      <bpmn:incoming>Flow_1ky5rfq</bpmn:incoming>
      <bpmn:outgoing>Flow_1e4331o</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id="Flow_1e4331o" sourceRef="Activity_1yfed52" targetRef="Activity_187vz7q" />
    <bpmn:exclusiveGateway id="Gateway_1a8lwpx" default="Flow_036mb80">
      <bpmn:incoming>Flow_0otfu2s</bpmn:incoming>
      <bpmn:outgoing>Flow_036mb80</bpmn:outgoing>
      <bpmn:outgoing>Flow_0ki7qj9</bpmn:outgoing>
    </bpmn:exclusiveGateway>
    <bpmn:sequenceFlow id="Flow_0otfu2s" sourceRef="Activity_187vz7q" targetRef="Gateway_1a8lwpx" />
    <bpmn:sequenceFlow id="Flow_036mb80" sourceRef="Gateway_1a8lwpx" targetRef="Activity_0qryqbu" />
    <bpmn:sequenceFlow id="Flow_0ki7qj9" name="UT2" sourceRef="Gateway_1a8lwpx" targetRef="Activity_0fcsn8g">
      <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression" language="JavaScript">(item.data.userCaptureCode=='UT2')</bpmn:conditionExpression>
    </bpmn:sequenceFlow>
    <bpmn:serviceTask id="Activity_0qryqbu" name="ST2" camunda:delegateExpression="Dummy">
      <bpmn:incoming>Flow_036mb80</bpmn:incoming>
      <bpmn:outgoing>Flow_05nb9oo</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id="Flow_0cahzxr" sourceRef="Activity_0fcsn8g" targetRef="Activity_03eqpfb" />
    <bpmn:endEvent id="Event_0d6l4hi" name="End">
      <bpmn:incoming>Flow_05nb9oo</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id="Flow_05nb9oo" sourceRef="Activity_0qryqbu" targetRef="Event_0d6l4hi" />
    <bpmn:endEvent id="Event_18f3u2i" name="End">
      <bpmn:incoming>Flow_1a1af1d</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id="Flow_1a1af1d" sourceRef="Activity_03eqpfb" targetRef="Event_18f3u2i" />
    <bpmn:startEvent id="Event_14l5yn1" name="RecoveryStart">
      <bpmn:extensionElements>
        <camunda:properties>
          <camunda:property name="source" value="MSNODE3" />
          <camunda:property name="protocol" value="H&#38;n" />
          <camunda:property name="modality" value="CT" />
        </camunda:properties>
      </bpmn:extensionElements>
      <bpmn:outgoing>Flow_0kqreig</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:sequenceFlow id="Flow_0kqreig" sourceRef="Event_14l5yn1" targetRef="Activity_09h88xv" />
    <bpmn:exclusiveGateway id="Gateway_1vlnqdl" default="Flow_05ur35v">
      <bpmn:incoming>Flow_04morto</bpmn:incoming>
      <bpmn:outgoing>Flow_05ur35v</bpmn:outgoing>
      <bpmn:outgoing>Flow_041el0a</bpmn:outgoing>
    </bpmn:exclusiveGateway>
    <bpmn:sequenceFlow id="Flow_04morto" sourceRef="Activity_09h88xv" targetRef="Gateway_1vlnqdl" />
    <bpmn:sequenceFlow id="Flow_05ur35v" sourceRef="Gateway_1vlnqdl" targetRef="Activity_187vz7q" />
    <bpmn:sequenceFlow id="Flow_041el0a" name="UT2" sourceRef="Gateway_1vlnqdl" targetRef="Activity_0fcsn8g">
      <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression" language="JavaScript">(item.data.userCaptureCode=='UT2')</bpmn:conditionExpression>
    </bpmn:sequenceFlow>
    <bpmn:serviceTask id="Activity_03eqpfb" name="ST3" camunda:delegateExpression="Dummy">
      <bpmn:incoming>Flow_0cahzxr</bpmn:incoming>
      <bpmn:outgoing>Flow_1a1af1d</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:userTask id="Activity_187vz7q" name="UT1">
      <bpmn:extensionElements>
        <camunda:inputOutput>
          <camunda:inputParameter name="userCaptureCode">hi</camunda:inputParameter>
        </camunda:inputOutput>
      </bpmn:extensionElements>
      <bpmn:incoming>Flow_1e4331o</bpmn:incoming>
      <bpmn:incoming>Flow_05ur35v</bpmn:incoming>
      <bpmn:outgoing>Flow_0otfu2s</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:userTask id="Activity_09h88xv" name="RecoveryUT">
      <bpmn:extensionElements>
        <camunda:inputOutput>
          <camunda:inputParameter name="userCaptureCode">UT2</camunda:inputParameter>
        </camunda:inputOutput>
      </bpmn:extensionElements>
      <bpmn:incoming>Flow_0kqreig</bpmn:incoming>
      <bpmn:outgoing>Flow_04morto</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:userTask id="Activity_0fcsn8g" name="UT2">
      <bpmn:incoming>Flow_0ki7qj9</bpmn:incoming>
      <bpmn:incoming>Flow_041el0a</bpmn:incoming>
      <bpmn:outgoing>Flow_0cahzxr</bpmn:outgoing>
    </bpmn:userTask>
  </bpmn:process>
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Process_1">
      <bpmndi:BPMNEdge id="Flow_041el0a_di" bpmnElement="Flow_041el0a">
        <di:waypoint x="495" y="250" />
        <di:waypoint x="710" y="250" />
        <di:waypoint x="710" y="160" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="592" y="232" width="21" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_05ur35v_di" bpmnElement="Flow_05ur35v">
        <di:waypoint x="470" y="225" />
        <di:waypoint x="470" y="160" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_04morto_di" bpmnElement="Flow_04morto">
        <di:waypoint x="360" y="250" />
        <di:waypoint x="445" y="250" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0kqreig_di" bpmnElement="Flow_0kqreig">
        <di:waypoint x="209" y="250" />
        <di:waypoint x="260" y="250" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1a1af1d_di" bpmnElement="Flow_1a1af1d">
        <di:waypoint x="910" y="120" />
        <di:waypoint x="962" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_05nb9oo_di" bpmnElement="Flow_05nb9oo">
        <di:waypoint x="770" y="10" />
        <di:waypoint x="952" y="10" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0cahzxr_di" bpmnElement="Flow_0cahzxr">
        <di:waypoint x="770" y="120" />
        <di:waypoint x="810" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0ki7qj9_di" bpmnElement="Flow_0ki7qj9">
        <di:waypoint x="605" y="120" />
        <di:waypoint x="670" y="120" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="628" y="102" width="21" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_036mb80_di" bpmnElement="Flow_036mb80">
        <di:waypoint x="580" y="95" />
        <di:waypoint x="580" y="10" />
        <di:waypoint x="670" y="10" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0otfu2s_di" bpmnElement="Flow_0otfu2s">
        <di:waypoint x="520" y="120" />
        <di:waypoint x="555" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1e4331o_di" bpmnElement="Flow_1e4331o">
        <di:waypoint x="360" y="120" />
        <di:waypoint x="420" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1ky5rfq_di" bpmnElement="Flow_1ky5rfq">
        <di:waypoint x="209" y="120" />
        <di:waypoint x="260" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="StartEvent_1">
        <dc:Bounds x="173" y="102" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="179" y="145" width="25" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_1c0rgal_di" bpmnElement="Activity_1yfed52">
        <dc:Bounds x="260" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Gateway_1a8lwpx_di" bpmnElement="Gateway_1a8lwpx" isMarkerVisible="true">
        <dc:Bounds x="555" y="95" width="50" height="50" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_1sgmyi0_di" bpmnElement="Activity_0qryqbu">
        <dc:Bounds x="670" y="-30" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_0d6l4hi_di" bpmnElement="Event_0d6l4hi">
        <dc:Bounds x="952" y="-8" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="960" y="35" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_18f3u2i_di" bpmnElement="Event_18f3u2i">
        <dc:Bounds x="962" y="102" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="970" y="145" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_14l5yn1_di" bpmnElement="Event_14l5yn1">
        <dc:Bounds x="173" y="232" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="156" y="275" width="71" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Gateway_1vlnqdl_di" bpmnElement="Gateway_1vlnqdl" isMarkerVisible="true">
        <dc:Bounds x="445" y="225" width="50" height="50" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0167smr_di" bpmnElement="Activity_03eqpfb">
        <dc:Bounds x="810" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0jrreoo_di" bpmnElement="Activity_187vz7q">
        <dc:Bounds x="420" y="80" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_1pg3c2f_di" bpmnElement="Activity_09h88xv">
        <dc:Bounds x="260" y="210" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_08wgr5c_di" bpmnElement="Activity_0fcsn8g">
        <dc:Bounds x="670" y="80" width="100" height="80" />
      </bpmndi:BPMNShape>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
  ```
</details>

**Comment by MaheshkumarSundaram** [comment]
@ralphhanna,

Thanks. I tried using the `src/scripts/cli.ts` in `bpmn-web` by adding `restart` option and it did work. 

**ISSUES:**

- Although the workflow has been restarted but the `status` is still at `end` instead of `running`. Shouldn't his be corrected?#
- Please update `bpmn-client` as the API needs to be changed accomodate optional params - `userName` & `options` for `startEvent`. The types are not included in it now which gives error when adding the parameters.

**Comment by MaheshkumarSundaram** [comment]
@ralphhanna,

Thanks. I tried using the `src/scripts/cli.ts` in `bpmn-web` by adding `restart` option and it did work. 

**ISSUES:**

- Although the workflow has been restarted, the `status` is still at `end` and `endedAt` has a date. Shouldn't `status` be `running` & `endedAt` be `null` when restarted?
- Please update `bpmn-client` as the API needs to be changed accomodate optional params - `userName` & `options` for `startEvent`. The types are not included in it now which gives error when adding the parameters.

**Comment by ralphhanna** [comment]
`bpmn-client` updated

**Comment by ralphhanna** [comment]
Yes
Sent from my iPhone

On Mar 21, 2024, at 10:37 AM, Maheshkumar Sundaram ***@***.***> wrote:

﻿

Hi @ralphhanna<https://github.com/ralphhanna>,

Although the workflow has been restarted, the status is still at end and endedAt has a date. Shouldn't status be running & endedAt be null when restarted?

Will this be sorted in the next release? Thanks

**Comment by ralphhanna** [comment]
Yes, working on it

**Comment by ralphhanna** [comment]
Please check new release 2.1.9

**Comment by ralphhanna** [comment]
I am going ahead and close this one as it is already completed 2.1.9

**Comment by MaheshkumarSundaram** [comment]
@ralphhanna,

This issue still seems to persist. I updated both `bpmn-web` and `bpmn-client` to the new versions.

1. `bpmn-client`'s `cli.ts` still throws error on `startEvent`.

    ![image](https://github.com/bpmnServer/bpmn-server/assets/55505611/0d81d312-a34c-4cec-a4b7-d10fdfd1c597)

   But the `cli.ts` under `src\scripts\` in `bpmn-web` doesn't give any error and it restarts the process.

2. Although `cli.ts` in `bpmn-web` restarts the process, the `status` of the process is still at `end` and `endedAt` has a date. I thought you fixed this too.

I have turned off my `enableSavePoints` but I suppose this doesn't affect this.

Kindly look into this. Thank you.

**Comment by ralphhanna** [comment]
Please get latest bpmn-server

**Comment by ralphhanna** [comment]
Also, the error message is valid, restart is only valid for completed processes.
`restart` has indeed been fixed so status is now running. however, `startEvent` is rather different and does not impact the status.

**Comment by MaheshkumarSundaram** [comment]
@ralphhanna,

`bpmn-server` is already updated but the `startEvent` from `bpmn-client` still gives error.

>>startEvent is rather different and does not impact the status.

May I ask why? This behaviour is similar to `restart` right? It doesn't seem logical if `status` and `endedAt` aren't updated!!!

**Comment by ralphhanna** [comment]
`startEvent` is intended for a second process to start to an already running instance
Example is 'invoice` below
![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/2ded7030-8068-420f-8753-f2e117950a55)

The second start allows the manager to intervene and cancel an invoice before it is completed or alter the behaviour in any way.
https://bpmnserver.github.io/bpmn-server/api/classes/APIEngine/#startevent

While `restart` is only used to **re-execute** an instance from a particular point due to an error, and this requires `savePoints` 
https://bpmnserver.github.io/bpmn-server/api/interfaces/IAPIEngine/#restart

**Comment by MaheshkumarSundaram** [comment]
@ralphhanna, 

Thanks for the explanation. But could you please clarify the below ambuiguity?

>> `startEvent` is intended for a second process to start to an already running instance

If it's intended to start an already running process, why `startEvent` has an option `{ restart: true }`? Isn't quite ambiguous as it means the instance has already completed and you are meaning to start a completed instance from a different start event?
If `{ restart: true }` is specified, it would make sense to update `status` and `endedAt` accordingly, and leave them unchanged if `restart` is `false` or not specified!!!?

What do you think?

**Comment by ralphhanna** [comment]
@MaheshkumarSundaram 
You win, my mistake, is fixed in the upcoming release 2.1.12
